src/Microsoft.Management.Infrastructure.CimCmdlets/CimAsyncOperation.cs

```diff
@@ -443,11 +442,14 @@ protected bool Disposed
         {
             get
             {
-                return Interlocked.Read(ref this._disposed) == 1;
+                // Ensure the read of _disposed doesn't move up before the write in the Dispose method.
```

Comment:
A memory barrier only guarantees the memory access order in the current thread, but `Disposed` method is very possible to be called in a different thread, so it doesn't seem useful to add the call to `Interlocked.MemoryBarrier()`.

Reply:
Yes it would seem we would need `Interlocked.MemoryBarrierProcessWide` instead, but according to documentation this is a very expensive operation.

The most correct method would be to use a dummy `CompareExchange` operation like in 6d1e96235 but @PaulHigin had concerns this would be using the Interlocked API in an unusual way.

It would be great if we could do `return Interlocked.Read(ref this._disposed) == 1;` but alas this is not defined for an Int32. I am considering opening a new issue at dotnet/runtime to add this API to the BCL.